### PR TITLE
Re-initialize static variables on startup/minit

### PR DESCRIPTION
**Pull request by morrisonlevi**
This is required because when Apache does a reload, the processes will
do a shutdown sequence but keep the process alive. It then goes through
startup again. Without re-initializing statics in startup they may be
stale and cause issues such as crashes.

Note that uv_once and pthread_once don't provide the necessary routines
to re-initialize a once, so I implemented a once component which allows
it, with restrictions the programmer will have to follow, of course.

### CODE REVIEW

File: components/once/once.c [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+void datadog_php_once(struct datadog_php_once_s *once,
+                      void (*init_routine)(void)) {
+  // If it's already initialized we can avoid the lock.
+  if (atomic_load(&once->initialized)) {
+    return;
+  }
+  uv_mutex_lock(&once->mutex);
+  // The value cannot be modified until the init routine has completed.
+  if (atomic_load(&once->initialized) == false) {
```
Comment (r1viollet) [inline components/once/once.c:25]: why do we need to re-test this value ?

Comment (r1viollet) [reply components/once/once.c:25]: Ah yes, if some other thread gets the lock between line 20 and line 23. Indeed.

File: profiling/datadog-profiling.c [hunk 2]
```diff
@@ -156,7 +172,23 @@ static void sapi_diagnose(sapi_t sapi, datadog_php_string_view pretty_name) {
   }
 }
 
-int datadog_profiling_startup(zend_extension *extension) {
+zend_result datadog_profiling_startup(zend_extension *extension) {
+  // Re-initialize static variables (Apache reload may have occurred)
+  int once_ctor_result = datadog_php_once_ctor(&first_activate_once);
```
Comment (r1viollet) [inline profiling/datadog-profiling.c:177]: We are never called by several threads here ?

File: components/once/once.c [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+#include <stdbool.h>
+#include <uv.h>
+
+int datadog_php_once_ctor(struct datadog_php_once_s *once) {
+  atomic_store(&once->initialized, false);
+  return uv_mutex_init(&once->mutex);
+}
+
+void datadog_php_once_dtor(struct datadog_php_once_s *once) {
+  atomic_store(&once->initialized, true);
```
Comment (r1viollet) [inline components/once/once.c:13]: Setting `true` here seems strange. It does guarantee that you will never call the init_routine again, though the boolean captures several states with this value.

File: components/once/once.c [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  uv_mutex_destroy(&once->mutex);
+}
+
+void datadog_php_once(struct datadog_php_once_s *once,
+                      void (*init_routine)(void)) {
+  // If it's already initialized we can avoid the lock.
+  if (atomic_load(&once->initialized)) {
+    return;
+  }
+  uv_mutex_lock(&once->mutex);
```
Comment (r1viollet) [inline components/once/once.c:23]: if you destroy while the lock is acquired, are things not going badly ?

File: components/once/once.c (hunk 3 above)
Comment (morrisonlevi) [reply components/once/once.c:13]: This is a destructor, so doing pretty much anything would be legitimate here for the var `initialized` -- if they use it again in any way except to `ctor` it, then it's invalid usage. Setting `initialized` to `true` in the destructor provides some amount of safety in the event of invalid usage, but that's a "best effort" kind of thing, I think?

Comment (r1viollet) [reply components/once/once.c:13]: I think it is OK, the only benefit of having a separate value (using an enum) is when you are investigating a possible crash and want to figure out which state things are in.

File: components/once/once.c (hunk 4 above)
Comment (morrisonlevi) [reply components/once/once.c:23]: Destructor and constructor are not thread-safe. I documented this in the header which will work fine for the intended use-case.

**Review by r1viollet — APPROVED** [review]
LGTM !

**Pull request merged**